### ISSUE TITLE
Testing options for orientation without mobile device

### ISSUE THREAD

**Issue by detlevhfischer** [w3c/wcag#643]
We have a question from evaluators with motor disabilities asking if there are ways to test 1.3.4 Orientation without having to rotate a mobile device. I have constructed a [simple demo page](http://3needs.org/en/testing/code/orientation.html) serving a warning based on media="screen and (orientation: landscape)" removing the warning with media="screen and (orientation: portrait)" http://3needs.org/en/testing/code/orientation.html
While the demo page demonstrates the failure (or one class of it) when displayed on a mobile device (tested on iPhone and Android (Nokia 6), I cannot emulate this test via the _sensors > orientation_ setting in Chrome along the [Google instructions](https://developers.google.com/web/tools/chrome-devtools/device-mode/orientation).  The nasty landscape orientation warning message does get displayed when in a desktop browser the viewport width is greater than the viewport height, but I hesitate to consider that equivalent to actually simulating the sensor input. Thoughts?
Is there another / a better way of doing it?

**Comment by patrickhlauke** [comment]
i'd be careful to distinguish here that *authors* are able do different things that would fail this SC:

* test the browser's width/height in JavaScript and infer if the browser is in portrait or landscape mode
* use CSS media queries (like `(orientation: landscape)`) to serve different styles (including for instance a style that just shows a "door slam" screen a la "please turn your device")
* use `matchMedia()` in JavaScript to run a CSS media query, as above
* read orientation data from the orientation sensor in JavaScript (`devicemotion` and `deviceorientation` events)

the first three can be tested by just changing your browser's aspect ratio even on desktop. the latter can only be tested with a real sensor or using the sensor orientation emulation.

what you were doing above, from what i understand, is trying to test for the first instance by using the approach needed to test the fourth instance.

**Comment by patrickhlauke** [comment]
also, to clarify: on a real device, what happens when you tilt the device between portrait and landscape isn't just that the sensor sends different data, but that the browser switches between portrait and landscape mode (unless the orientation lock is enabled at OS level, or the page has explicitly requested that the browser not do this using the [Device Adaptation `orientation` descriptor](https://drafts.csswg.org/css-device-adapt/#orientation-desc), where supported). that's the thing that just simulating the sensor in Chrome Devtools doesn't do.

**Comment by detlevhfischer** [comment]
Ah ok, thanks, Patrick. So then it seems that to catch the cases where the `deviceorientation` event is used, one would have to use a mobile device for testing (i.e. a motor-impaired tester would need to get someone else to check that for them).

**Comment by detlevhfischer** [comment]
...and I guess detecting the presence of sensor events will run into the same problem as trying to detect down events when these are attached via event listeners, correct? We have been made aware of a [Visual Event bookmarklet](http://www.sprymedia.co.uk/VisualEvent/) ( http://www.sprymedia.co.uk/VisualEvent/ ) which **may** turn out to be useful here, but I haven't evaluated that yet.

**Comment by mraccess77** [comment]
In my experience event detection is often not possible directly from the DOM when events are added unless you can capture the events before the page has loaded, events are inline attributes, you are using a browser extension or if your page uses something like Jquery were the events are added through a framework that track them.

**Comment by patrickhlauke** [comment]
> Ah ok, thanks, Patrick. So then it seems that to catch the cases where the `deviceorientation` event is used, one would have to use a mobile device for testing (i.e. motor-impaired testers would need to get someone else to check that for them).

no, you can catch those using either a real device OR a simulator/emulation that allows for these events to be fired (like the thing in the chrome devtools that you mention in the thread starter).

**Comment by patrickhlauke** [comment]
> ...and I guess detecting the presence of sensor events will run into the same problem as trying to detect down events when these are attached via event listeners, correct? We have been made aware of a [Visual Event bookmarklet](http://www.sprymedia.co.uk/VisualEvent/) ( http://www.sprymedia.co.uk/VisualEvent/ ) which **may** turn out to be useful here, but I haven't evaluated that yet.

You'd either want to test with an actual device (same as for the down event thing where you'd really want to functionally test), OR see if you can at least do some static analysis of all JavaScript and look for telltale signs of relevant even listener names (e.g. doing a very naive "find" across all JS files for, say, `deviceorientation` - and `pointerdown`, `mousedown`, `touchstart` ...) to at least get a first indication

**Comment by fstrr** [comment]
This discussion seems to have become inactive, so I'm going to close it. If it needs to be re-opened, please do.
